rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,641 @@
+Even if increasing the number of workers doesn’t just linearly increase the performances 
+(which totally make sense, most of our requests response time is caused by awaiting ES response), 
+the improvements of the clustering mode on performance under heavy load are far from negligible.
+
+## 4.2 Testing against cloud
+
+There is currently no easy way to test the performance improvements this could provide on Cloud, as we can't
+deploy custom builds or branches on Cloud at the moment.
+
+On Cloud, Kibana is running in a containerised environment using CPU CFS quota and CPU shares. 
+
+If we want to investigate the potential perf improvement on Cloud further, our only option would be to setup a 
+similar-ish environment locally (which wasn't done during the initial investigation).
+
+# 5. Detailed design
+
+## 5.1 Enabling clustering mode
+
+Enabling clustering mode will be done using the `clustering.enabled` configuration property.
+
+If clustering is enabled by default, then no configuration would be required by users, and
+Kibana would automatically use all available cores. However, more detailed configuration
+would be available for users with more advanced use cases:
+```yaml
+clustering:
+  enabled: true               # enabled by default
+  coordinator:
+    max_old_space_size: 1gb   # optional, allows to configure memory limit for coordinator only
+  workers:                    # when provided, total # of workers would be based on # of items in this list
+    - name: foo               # not technically necessary, but useful for logging purposes
+      max_old_space_size: 1gb # optional, allows to configure memory limits per-worker
+    - name: bar
+      max_old_space_size: 1gb
+```
+
+This per-worker design would give us the flexibility to eventually provide more fine-grained configuration,
+like dedicated workers for http requests or background jobs.
+
+## 5.2 Cross-worker communication
+
+For some of our changes (such as the `/status` API, see below), we will need some kind of cross-worker
+communication. This will need to pass through the coordinator, which will also serve as an 'event bus',
+or IPC forwarder.
+
+This IPC API will be exposed from the clustering service:
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
+  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
+}
+```
+
+To preserve isolation and to avoid creating an implicit cross-plugin API, handlers registered from a
+given plugin will only be invoked for messages sent by the same plugin.
+
+Notes:
+-  To reduce clustered and non-clustered mode divergence, in non-clustered mode, these APIs would just be no-ops. 
+   It will avoid forcing (most) code to check which mode Kibana is running before calling them.
+-  We could eventually use an Observable pattern instead of a handler pattern to subscribe to messages.
```

Comment:
Still a little uneasy about not using Observables here. This would be one of the few (only?) Core event APIs that does not use Observables, however my pains with memory and performance overhead with Observables on the status API does make me a bit less cautious about this divergence here.

Reply:
Yeah perf overhead was my main concern here. I guess the options are:
- We go with the existing proposal, evaluate how it's going, and decide whether to add observables as an enhancement later
- We drop the existing proposal and go with observables from the beginning, and see if we can track memory usage to determine whether this becomes a problem down the road

Thoughts? Anyone have strong feelings on this?

I think observables is more optimal from a design perspective, but the existing proposal feels less risky.

Reply:
See https://github.com/elastic/kibana/pull/94057#discussion_r620028092 for more details about the complexity of the observable based API

> I think observables is more optimal from a design perspective, but the existing proposal feels less risky.

My thoughts exactly. Would love to have an obs API, but implementation seems (way?) more complex to handle all the edge cases

Reply:
Yep, I'm fine with this for now. We explore an Observable implementation later and evaluate the cost. Consistency in the Core APIs is really valuable IMO, but stability and performance takes precedent.